src/main/java/com/google/devtools/build/lib/skyframe/TreeArtifactValue.java

```diff
@@ -190,23 +192,35 @@ public static ArchivedRepresentation create(
    */
   @Nullable private final ArchivedRepresentation archivedRepresentation;
 
+  /**
+   * Optional symlink target path.
+   *
+   * <p>If set, this artifact is a copy of another artifact. It is materialized in the local
+   * filesystem as a symlink to the original artifact, whose contents live at this location. This is
+   * used to implement zero-cost copies of remotely stored artifacts.
+   */
+  @Nullable private final PathFragment symlinkTargetExecPath;
+
   private final boolean entirelyRemote;
 
   private TreeArtifactValue(
       byte[] digest,
       ImmutableSortedMap<TreeFileArtifact, FileArtifactValue> childData,
       long totalChildSize,
       @Nullable ArchivedRepresentation archivedRepresentation,
+      @Nullable PathFragment symlinkTargetExecPath,
       boolean entirelyRemote) {
     this.digest = digest;
     this.childData = childData;
     this.totalChildSize = totalChildSize;
     this.archivedRepresentation = archivedRepresentation;
+    this.symlinkTargetExecPath = symlinkTargetExecPath;
     this.entirelyRemote = entirelyRemote;
   }
 
   public FileArtifactValue getMetadata() {
-    return FileArtifactValue.createProxy(digest);
+    return FileArtifactValue.createProxyForTreeArtifact(
```

Comment:
Given you have a public method for the symlink execpath, I wonder if that would not better be a private static class in here or even an anonymous class here. This is closely tied to the tree artifact, yet the code lives quite far from each other.

Reply:
Done in 5faf917.